closed/adds/jdk/src/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    @Override
+    protected void engineInit(Key key, SecureRandom random)
+            throws InvalidKeyException {
+        if (!(key instanceof PrivateKey)) {
+            throw new InvalidKeyException
+                ("Key must be an instance of PrivateKey");
+        }
+        // attempt to translate the key if it is not an ECKey
+        this.privateKey = (ECPrivateKeyImpl) ECKeyFactory.toECKey(key);
+        this.publicKey = null;
+    }
+
+    @Override
+    protected void engineInit(Key key, AlgorithmParameterSpec params, SecureRandom random)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (params != null) {
+            throw new InvalidAlgorithmParameterException
+                        ("Parameters not supported");
+        }
+        engineInit(key, random);
+    }
+
+    @Override
+    protected Key engineDoPhase(Key key, boolean lastPhase)
+            throws InvalidKeyException, IllegalStateException {
+        if (this.privateKey == null) {
+            throw new IllegalStateException("Not initialized");
+        }
+        if (this.publicKey != null) {
+            throw new IllegalStateException("Phase already executed");
+        }
+        if (!lastPhase) {
+            throw new IllegalStateException
+                ("Only two party agreement supported, lastPhase must be true");
+        }
+        if (!(key instanceof PublicKey)) {
+            throw new InvalidKeyException
+                ("Key must be an instance of PublicKey");
+        }
+        // attempt to translate the key if it is not an ECKey
+        this.publicKey = (ECPublicKeyImpl) ECKeyFactory.toECKey(key);
+
+        ECParameterSpec params = this.publicKey.getParams();
+        int keyLenBits = params.getCurve().getField().getFieldSize();
+        this.secretLen = (keyLenBits + 7) >> 3;
+
+        return null;
+    }
+
+    @Override
+    protected byte[] engineGenerateSecret() throws IllegalStateException {
+        byte[] secret = new byte[this.secretLen];
+        try {
+            engineGenerateSecret(secret, 0);
+        } catch (ShortBufferException e) {
+            /* should not happen */
+        }
+        return secret;
+    }
+
+    @Override
+    protected int engineGenerateSecret(byte[] sharedSecret, int offset)
+            throws IllegalStateException, ShortBufferException {
+        if ((offset + this.secretLen) > sharedSecret.length) {
+            throw new ShortBufferException("Need " + this.secretLen
+                + " bytes, only " + (sharedSecret.length - offset)
+                + " available");
+        }
+        if ((this.privateKey == null) || (this.publicKey == null)) {
+            throw new IllegalStateException("Not initialized correctly");
+        }
+        long nativePublicKey = this.publicKey.getNativePtr();
+        long nativePrivateKey = this.privateKey.getNativePtr();
+        if ((nativePublicKey < 0) || (nativePrivateKey < 0)) {
```

Comment:
These should check for `-1`, rather than any negative number; see the javadoc for `getNativePtr()`.